File: backend/app/api/routes/roster.py
import re
import uuid

from fastapi import APIRouter, HTTPException, status
from sqlalchemy import select
from sqlalchemy.orm import selectinload

from app.api.deps import CurrentUser, DbSession
from app.core.security import hash_password
from app.models.course import Enrollment
from app.models.user import Discipline, Role, User
from app.schemas.clinical import (
    RosterImportIn,
    RosterImportOut,
    RosterMemberIn,
    RosterMemberOut,
    SessionUserOut,
)
from app.services.access import can_manage_roster, list_enrollments
from app.services.audit import log_event
from app.services.serializers import session_user_from

router = APIRouter(prefix="/courses/{course_id}/roster", tags=["roster"])

TEMP_PASSWORD = "ChangeMe1!"
ALLOWED_EMAIL = ("@utep.edu", "@miners.utep.edu")
UNI_RE = re.compile(r"^800\d{6}$")


def _assert_utep_email(email: str) -> str:
    lowered = email.strip().lower()
    if not any(lowered.endswith(sfx) for sfx in ALLOWED_EMAIL):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Use a UTEP email (@utep.edu or @miners.utep.edu).",
        )
    return lowered


def _parse_name(full_name: str) -> tuple[str, str]:
    parts = full_name.strip().split(None, 1)
    first = parts[0] if parts else "Student"
    last = parts[1] if len(parts) > 1 else ""
    return first, last


async def _role_by_code(db: DbSession, code: str) -> Role | None:
    return (
        await db.execute(select(Role).options(selectinload(Role.permissions)).where(Role.code == code))
    ).scalar_one_or_none()


@router.get("", response_model=list[SessionUserOut])
async def list_roster(course_id: uuid.UUID, user: CurrentUser, db: DbSession):
    if not can_manage_roster(user):
        raise HTTPException(status_code=403, detail="Not allowed to view roster")
    rows = (
        await db.execute(
            select(Enrollment).where(
                Enrollment.course_id == course_id,
                Enrollment.app_role.in_(("student", "instructor")),
            )
        )
    ).scalars().all()
    out: list[SessionUserOut] = []
    seen: set[uuid.UUID] = set()
    for e in rows:
        if e.user_id in seen:
            continue
        seen.add(e.user_id)
        u = await db.get(
            User,
            e.user_id,
            options=(selectinload(User.roles).selectinload(Role.permissions), selectinload(User.discipline)),
        )
        if u is None:
            continue
        enrollments = await list_enrollments(db, u.id)
        out.append(session_user_from(u, enrollments))
    return out


@router.post("/members", response_model=RosterMemberOut, status_code=status.HTTP_201_CREATED)
async def add_member(course_id: uuid.UUID, body: RosterMemberIn, user: CurrentUser, db: DbSession):
    """Add one student or instructor to the course (requires user:manage)."""
    if not can_manage_roster(user):
        raise HTTPException(status_code=403, detail="Not allowed to add users")

    email = _assert_utep_email(str(body.email))
    uni = (body.universityId or "").strip() or None
    if uni and not UNI_RE.match(uni):
        raise HTTPException(status_code=400, detail="800 number must look like 800123456.")

    app_role = body.appRole
    if app_role not in ("student", "instructor"):
        raise HTTPException(status_code=400, detail="Role must be student or instructor.")

    role = await _role_by_code(db, app_role)
    discipline = (
        await db.execute(select(Discipline).where(Discipline.code == body.discipline))
    ).scalar_one_or_none()
    if app_role == "student" and discipline is None:
        raise HTTPException(status_code=400, detail="Unknown discipline.")

    first, last = _parse_name(body.fullName)
    existing = (
        await db.execute(
            select(User)
            .options(selectinload(User.roles).selectinload(Role.permissions), selectinload(User.discipline))
            .where(User.email == email)
        )
    ).scalar_one_or_none()
    created = False
    if existing is None:
        existing = User(
            email=email,
            first_name=first,
            last_name=last,
            hashed_password=hash_password(TEMP_PASSWORD),
            must_change_password=True,
            university_id=uni,
            discipline_id=discipline.id if discipline else None,
            roles=[role] if role else [],
        )
        db.add(existing)
        await db.flush()
        created = True
    else:
        if uni and not existing.university_id:
            existing.university_id = uni
        if role and all(r.id != role.id for r in existing.roles):
            existing.roles.append(role)

    disc_code = body.discipline if app_role == "student" else None
    enr_q = select(Enrollment).where(
        Enrollment.user_id == existing.id,
        Enrollment.course_id == course_id,
        Enrollment.app_role == app_role,
    )
    if disc_code is None:
        enr_q = enr_q.where(Enrollment.discipline_code.is_(None))
    else:
        enr_q = enr_q.where(Enrollment.discipline_code == disc_code)
    enr = (await db.execute(enr_q)).scalar_one_or_none()
    already = enr is not None
    if not already:
        db.add(
            Enrollment(
                user_id=existing.id,
                course_id=course_id,
                app_role=app_role,
                discipline_code=disc_code,
            )
        )

    await log_event(
        db,
        action="roster.add_member",
        entity_type="course",
        actor_user_id=user.id,
        entity_id=str(course_id),
        details={"email": email, "appRole": app_role, "created": created, "alreadyEnrolled": already},
    )
    await db.commit()

    enrollments = await list_enrollments(db, existing.id)
    # reload with relationships
    existing = await db.get(
        User,
        existing.id,
        options=(selectinload(User.roles).selectinload(Role.permissions), selectinload(User.discipline)),
    )
    assert existing is not None
    return RosterMemberOut(
        created=created,
        alreadyEnrolled=already,
        user=session_user_from(existing, enrollments),
    )


@router.post("/import", response_model=RosterImportOut)
async def import_roster(course_id: uuid.UUID, body: RosterImportIn, user: CurrentUser, db: DbSession):
    if not can_manage_roster(user):
        raise HTTPException(status_code=403, detail="Not allowed to import roster")
    if any(r.get("problem") for r in body.rows):
        raise HTTPException(status_code=400, detail="Fix the flagged rows before importing.")

    student_role = await _role_by_code(db, "student")
    discipline = (
        await db.execute(select(Discipline).where(Discipline.code == body.discipline))
    ).scalar_one_or_none()

    added = 0
    already = 0
    for row in body.rows:
        email = row["email"].strip().lower()
        full_name = row["fullName"].strip()
        university_id = row.get("universityId", "").strip() or None
        first, last = _parse_name(full_name)

        existing = (await db.execute(select(User).where(User.email == email))).scalar_one_or_none()
        if existing is None:
            existing = User(
                email=email,
                first_name=first,
                last_name=last,
                hashed_password=hash_password(TEMP_PASSWORD),
                must_change_password=True,
                university_id=university_id,
                discipline_id=discipline.id if discipline else None,
                roles=[student_role] if student_role else [],
            )
            db.add(existing)
            await db.flush()

        enr = (
            await db.execute(
                select(Enrollment).where(
                    Enrollment.user_id == existing.id,
                    Enrollment.course_id == course_id,
                    Enrollment.app_role == "student",
                    Enrollment.discipline_code == body.discipline,
                )
            )
        ).scalar_one_or_none()
        if enr:
            already += 1
        else:
            db.add(
                Enrollment(
                    user_id=existing.id,
                    course_id=course_id,
                    app_role="student",
                    discipline_code=body.discipline,
                )
            )
            added += 1

    await log_event(
        db,
        action="roster.import",
        entity_type="course",
        actor_user_id=user.id,
        entity_id=str(course_id),
        details={"added": added, "alreadyEnrolled": already},
    )
    await db.commit()
    return RosterImportOut(added=added, alreadyEnrolled=already)


@router.delete("/{user_id}", status_code=status.HTTP_204_NO_CONTENT)
async def remove_from_course(course_id: uuid.UUID, user_id: uuid.UUID, user: CurrentUser, db: DbSession):
    if not can_manage_roster(user):
        raise HTTPException(status_code=403, detail="Not allowed")
    rows = (
        await db.execute(
            select(Enrollment).where(
                Enrollment.course_id == course_id,
                Enrollment.user_id == user_id,
            )
        )
    ).scalars().all()
    for r in rows:
        await db.delete(r)
    await log_event(
        db,
        action="roster.remove",
        entity_type="course",
        actor_user_id=user.id,
        entity_id=str(course_id),
        details={"user": str(user_id)},
    )
    await db.commit()

File: frontend/src/api/roster.ts
import type { Discipline, RosterRow, User } from "../types";
import { ApiError, USE_MOCK, apiFetch, clone, latency, uid } from "./client";
import { db } from "./mockDb";
import { recordAudit } from "./audit";

export async function listRoster(courseId: string): Promise<User[]> {
  if (!USE_MOCK) return apiFetch<User[]>(`/courses/${courseId}/roster`);
  await latency(150);
  return clone(
    db.users.filter((u) =>
      u.roles.some((r) => (r.role === "student" || r.role === "instructor") && r.courseIds.includes(courseId)),
    ),
  );
}

export async function addRosterMember(
  actor: User,
  courseId: string,
  input: {
    fullName: string;
    email: string;
    universityId?: string;
    discipline: Discipline;
    appRole: "student" | "instructor";
  },
): Promise<{ created: boolean; alreadyEnrolled: boolean; user: User }> {
  if (!USE_MOCK) {
    return apiFetch<{ created: boolean; alreadyEnrolled: boolean; user: User }>(
      `/courses/${courseId}/roster/members`,
      {
        method: "POST",
        body: {
          fullName: input.fullName.trim(),
          email: input.email.trim().toLowerCase(),
          universityId: input.universityId?.trim() || undefined,
          discipline: input.discipline,
          appRole: input.appRole,
        },
      },
    );
  }
  await latency(250);
  const email = input.email.trim().toLowerCase();
  if (!/@(miners\.)?utep\.edu$/i.test(email)) {
    throw new ApiError(400, "Use a UTEP email (@utep.edu or @miners.utep.edu).");
  }
  let user = db.users.find((u) => u.email.toLowerCase() === email);
  let created = false;
  if (!user) {
    user = {
      id: uid("u"),
      fullName: input.fullName.trim(),
      email,
      universityId: input.universityId?.trim() || undefined,
      roles: [],
      mustChangePassword: true,
    };
    db.users.push(user);
    created = true;
  }
  let assignment = user.roles.find(
    (r) => r.role === input.appRole && (input.appRole === "instructor" || r.discipline === input.discipline),
  );
  if (!assignment) {
    assignment = {
      role: input.appRole,
      discipline: input.appRole === "student" ? input.discipline : undefined,
      courseIds: [],
    };
    user.roles.push(assignment);
  }
  let alreadyEnrolled = false;
  if (assignment.courseIds.includes(courseId)) {
    alreadyEnrolled = true;
  } else {
    assignment.courseIds.push(courseId);
  }
  recordAudit(actor, "roster.add_member", `course/${courseId}`, "ok", `${email} as ${input.appRole}`);
  return { created, alreadyEnrolled, user: clone(user) };
}

export async function importRoster(
  actor: User, courseId: string, discipline: Discipline, rows: RosterRow[],
): Promise<{ added: number; alreadyEnrolled: number }> {
  if (!USE_MOCK) {
    return apiFetch<{ added: number; alreadyEnrolled: number }>(`/courses/${courseId}/roster/import`, {
      method: "POST",
      body: {
        discipline,
        rows: rows.map((r) => ({
          fullName: r.fullName,
          email: r.email,
          universityId: r.universityId,
          ...(r.problem ? { problem: r.problem } : {}),
        })),
      },
    });
  }
  await latency(400);
  if (rows.some((r) => r.problem)) throw new ApiError(400, "Fix the flagged rows before importing.");
  let added = 0;
  let alreadyEnrolled = 0;
  for (const row of rows) {
    let user = db.users.find((u) => u.email.toLowerCase() === row.email.toLowerCase());
    if (!user) {
      user = { id: uid("u"), fullName: row.fullName, email: row.email, universityId: row.universityId, roles: [] };
      db.users.push(user);
    }
    let assignment = user.roles.find((r) => r.role === "student" && r.discipline === discipline);
    if (!assignment) {
      assignment = { role: "student", discipline, courseIds: [] };
      user.roles.push(assignment);
    }
    if (assignment.courseIds.includes(courseId)) {
      alreadyEnrolled++;
    } else {
      assignment.courseIds.push(courseId);
      added++;
    }
  }
  recordAudit(actor, "roster.import", `course/${courseId}`, "ok", `${added} added, ${alreadyEnrolled} already enrolled`);
  return { added, alreadyEnrolled };
}

export async function removeFromCourse(actor: User, courseId: string, userId: string): Promise<void> {
  if (!USE_MOCK) {
    await apiFetch<void>(`/courses/${courseId}/roster/${userId}`, { method: "DELETE" });
    return;
  }
  await latency(200);
  const user = db.users.find((u) => u.id === userId);
  user?.roles.forEach((r) => {
    if (r.role === "student" || r.role === "instructor") {
      r.courseIds = r.courseIds.filter((c) => c !== courseId);
    }
  });
  recordAudit(actor, "roster.remove", `course/${courseId}`, "ok", `user/${userId}`);
}
